modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -162,27 +206,62 @@ public CompletableFuture<Void> cleanup(
         return allOf(cleanupFutures);
     }
 
-    /** {@inheritDoc} */
     @Override
     public int finished() {
         return (int) states.entrySet().stream().filter(e -> e.getValue() == TxState.COMMITED || e.getValue() == TxState.ABORTED).count();
     }
 
-    /** {@inheritDoc} */
     @Override
     public void start() {
         // No-op.
     }
 
-    /** {@inheritDoc} */
     @Override
-    public void stop() throws Exception {
+    public void stop() {
         // No-op.
     }
 
-    /** {@inheritDoc} */
     @Override
     public LockManager lockManager() {
         return lockManager;
     }
+
+    CompletableFuture<Void> completeReadOnlyTransactionFuture(ReadOnlyTxId readOnlyTxId) {
+        CompletableFuture<Void> readOnlyTxFuture = readOnlyTxFutureById.remove(readOnlyTxId);
+
+        assert readOnlyTxFuture != null : readOnlyTxId;
+
+        readOnlyTxFuture.complete(null);
+
+        return readOnlyTxFuture;
+    }
+
+    @Override
+    public void updateLowWatermark(HybridTimestamp newLowWatermark) {
+        lowWatermarkReadWriteLock.writeLock().lock();
+
+        try {
+            lowWatermark.updateAndGet(previousLowWatermark -> {
+                if (previousLowWatermark == null) {
+                    return newLowWatermark;
+                }
+
+                assert newLowWatermark.compareTo(previousLowWatermark) > 0 :
+                        "lower watermark should be growing: [previous=" + previousLowWatermark + ", new=" + newLowWatermark + ']';
+
+                return newLowWatermark;
+            });
+        } finally {
+            lowWatermarkReadWriteLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> getFutureAllReadOnlyTransactionsWhichLessOrEqualTo(HybridTimestamp timestamp) {
+        ReadOnlyTxId upperBound = new ReadOnlyTxId(timestamp, new UUID(Long.MAX_VALUE, Long.MAX_VALUE));
+
+        List<CompletableFuture<Void>> readOnlyTxFutures = List.copyOf(readOnlyTxFutureById.headMap(upperBound, true).values());
```

Comment:
I missed this first time, but: this method will only work correctly if the `timestamp` that was passed is <= than the current LWM. Could you please add the corresponding assertion and mention this in the javadoc?

Reply:
I don't think it's necessary, this method is more like a general case.

Reply:
Discussed in person, decided to change the signature of method `org.apache.ignite.internal.tx.TxManager#updateLowWatermark`.